src/ORM/Association/HasMany.php

```diff
@@ -134,6 +181,63 @@ public function saveAssociated(EntityInterface $entity, array $options = [])
     }
 
     /**
+     * Deletes/sets null the related objects according to the dependency between source and targets and foreign key nullability
+     * Skips deleting records present in $remainingEntities
+     *
+     * @param array $properties array of foreignKey properties
+     * @param EntityInterface $entity the entity which should have its associated entities unassigned
+     * @param Table $target The associated table
+     * @param array $remainingEntities Entities that should not be deleted
+     * @return void
+     */
+    protected function _unlinkAssociated(array $properties, EntityInterface $entity, Table $target, array $remainingEntities)
+    {
+        $primaryKey = (array)$target->primaryKey();
+        $mustBeDependent = (!$this->_foreignKeyAcceptsNull($target, $properties) || $this->dependent());
+        $conditions = [
+            'AND' => [
+                'NOT' => [
+                    'AND' => array_map(
+                        function ($ent) use ($primaryKey) {
+                            return $ent->extract($primaryKey);
+                        },
+                        $remainingEntities
+                    )
+                ],
+                $properties
+            ]
+        ];
+
+        $this->dependent($mustBeDependent);
+
+        if ($mustBeDependent) {
+                $target->deleteAll($conditions);
+        } else {
+            $updateFields = array_fill_keys(array_keys($properties), null);
+            $target->updateAll($updateFields, $conditions);
+        }
+    }
+
+    /**
+     * Checks the nullable flag of the foreign key
+     *
+     * @param Table $table the table containing the foreign key
+     * @param array $properties the list of fields that compose the foreign key
+     * @return bool
+     */
+    protected function _foreignKeyAcceptsNull(Table $table, array $properties)
+    {
+        return (bool)array_product(
```

Comment:
Using array_product() on a list of booleans is a bit tricksey. Couldn't you get the same results with array_filter() and empty()?